sys/net/gnrc/application_layer/tftp/gnrc_tftp.c

```diff
@@ -0,0 +1,848 @@
+
+    while (1) {
+        int ret = BUSY;
+        bool got_client = false;
+
+        /* register our TFTP response listener */
+        if (gnrc_netreg_register(GNRC_NETTYPE_UDP, &entry)) {
+            DEBUG("tftp: error starting server.");
+            return FAILED;
+        }
+
+        /* main processing loop */
+        while (ret == BUSY) {
+            /* wait for a message */
+            msg_receive(&msg);
+            ret = _tftp_state_processes(ctxt, &msg);
+
+            if (ret == BUSY && !got_client) {
+                gnrc_netreg_unregister(GNRC_NETTYPE_UDP, &entry);
+                got_client = true;
+            }
+        }
+    }
+
+    /* unregister our UDP listener on this thread */
+    gnrc_netreg_unregister(GNRC_NETTYPE_UDP, &entry);
+
+    return 0;
+}
+
+int _tftp_do_client_transfer(tftp_context_t *ctxt) {
+    msg_t msg;
+    tftp_state ret = BUSY;
+
+    /* register our DNS response listener */
+    gnrc_netreg_entry_t entry = { NULL, ctxt->src_port, thread_getpid() };
+    if (gnrc_netreg_register(GNRC_NETTYPE_UDP, &entry)) {
+        DEBUG("tftp: error starting server.");
+        return FAILED;
+    }
+
+    /* try to start the TFTP transfer */
+    ret = _tftp_state_processes(ctxt, NULL);
+    if (ret != BUSY) {
+        /* if the start failed return */
+        return ret;
+    }
+
+    /* main processing loop */
+    while (ret == BUSY) {
+        /* wait for a message */
+        msg_receive(&msg);
+        ret = _tftp_state_processes(ctxt, &msg);
+    }
+
+    /* unregister our UDP listener on this thread */
+    gnrc_netreg_unregister(GNRC_NETTYPE_UDP, &entry);
+
+    return ret;
+}
+
+tftp_state _tftp_state_processes(tftp_context_t *ctxt, msg_t *m) {
+    gnrc_pktsnip_t *outbuf = gnrc_pktbuf_add(NULL, NULL, TFTP_DEFAULT_DATA_SIZE,
+                                             GNRC_NETTYPE_UNDEF);
+
+    /* check if this is an client start */
+    if (!m) {
+        return _tftp_send_start(ctxt, outbuf);
+    }
+    else if (m->type == TFTP_TIMEOUT_MSG) {
+        if (++(ctxt->retries) > GNRC_TFTP_MAX_RETRIES) {
+            /* transfer failed due to lost peer */
+            return FAILED;
+        }
+
+        /* increase the timeout for congestion control */
+        ctxt->block_timeout <<= 1;
+
+        /* the send message timed out, re-sending */
+        if (ctxt->dst_port == GNRC_TFTP_DEFAULT_DST_PORT) {
+            /* we are still negotiating resent, start */
+            return _tftp_send_start(ctxt, outbuf);
+        }
+        else {
+            /* we are sending / receiving data */
+            /* if we are reading resent the ACK, if writing the DATA */
+            return _tftp_send_dack(ctxt, outbuf, (ctxt->op == TO_RRQ) ? TO_ACK : TO_DATA);
+        }
+    }
+    else if (m->type != GNRC_NETAPI_MSG_TYPE_RCV) {
+        DEBUG("tftp: unknown message");
+        return BUSY;
+    }
+
+    gnrc_pktsnip_t *pkt = (gnrc_pktsnip_t*)(m->content.ptr);
```

Comment:
If I'm not mistaken `pkt` needs to be released just like `outbuf`, see other comments.